src/Symfony/Component/Form/Extension/Core/Type/CurrencyType.php

```diff
@@ -11,36 +11,42 @@
 
 namespace Symfony\Component\Form\Extension\Core\Type;
 
-use Symfony\Component\Form\AbstractType;
 use Symfony\Component\Intl\Intl;
-use Symfony\Component\OptionsResolver\OptionsResolver;
 
-class CurrencyType extends AbstractType
+class CurrencyType extends ImmutableChoiceType
 {
     /**
-     * {@inheritdoc}
+     * Stores the available currency choices.
      */
-    public function configureOptions(OptionsResolver $resolver)
-    {
-        $resolver->setDefaults(array(
-            'choices' => array_flip(Intl::getCurrencyBundle()->getCurrencyNames()),
-            'choice_translation_domain' => false,
-        ));
-    }
+    private static $currencies;
 
     /**
      * {@inheritdoc}
      */
-    public function getParent()
+    public function getBlockPrefix()
     {
-        return __NAMESPACE__.'\ChoiceType';
+        return 'currency';
     }
 
     /**
      * {@inheritdoc}
+     *
+     * Returns the currency choices.
+     *
+     * The choices are generated from the Intl component
+     * \Symfony\Component\Intl\Intl::getCurrencyBundle().
+     * They are cached during a single request, so multiple
+     * currency fields on the same page don't lead to
+     * unnecessary overhead.
+     *
+     * @return array The currency choices
      */
-    public function getBlockPrefix()
+    protected function loadChoices()
     {
-        return 'currency';
+        if (null === self::$currencies) {
```

Comment:
@stof this PR is now about the caching in a first commit and a refactoring in a second commit.

For `choice_filter` and a better approach in the implementation see #18334.

I would very much appreciate some feedback there before adding more tests. Thanks.